flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/collector/TableFunctionResultFuture.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.collector;
+
+import org.apache.flink.api.common.functions.AbstractRichFunction;
+import org.apache.flink.streaming.api.functions.async.ResultFuture;
+
+/**
+ * The basic implementation of collector for {@link ResultFuture} in table joining.
+ */
+public abstract class TableFunctionResultFuture<T> extends AbstractRichFunction implements ResultFuture<T> {
```

Comment:
The responsibility of this class and `TableFunctionCollector` is not clean. It looks like they want to do some encapsulation but actually only contain some get and set methods. And the interface contract also seems not consistent, see the comment i left in `TableFunctionCollector`

Reply:
The responsibility of them are used to generate a collector which will do some filters or projections on both left input row and right table row. That's why we have `setInput`, `getInput` and a `setCollector` to set the real underlying collector.

The only difference between them is `TableFunctionCollector` has an additional `collected` flag used to indicate whether right table is empty. Because UDTF may call `collect(T)` zero or several times, we need a way to know whether it is called zero times, so that we can emit an null row for left join. However, `ResultFuture.complete(Collection<OUT> result)` will be called exactly once, so that we don't need a flag to indicate the "zero call", an empty or null `result` is the "zero call".